src/devices/uart.c

```diff
@@ -0,0 +1,160 @@
+/*
+ * rv32emu is freely redistributable under the MIT License. See the file
+ * "LICENSE" for information on usage and redistribution of this file.
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <poll.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "uart.h"
+
+/* Emulate 8250 (plain, without loopback mode support) */
+
+#define U8250_INT_THRE 1
+
+void u8250_update_interrupts(u8250_state_t *uart)
+{
+    /* Some interrupts are level-generated. */
+    /* TODO: does it also generate an LSR change interrupt? */
+    if (uart->in_ready)
+        uart->pending_intrs |= 1;
+    else
+        uart->pending_intrs &= ~1;
+
+    /* Prevent generating any disabled interrupts in the first place */
+    uart->pending_intrs &= uart->ier;
+
+    /* Update current interrupt (higher bits -> more priority) */
+    if (uart->pending_intrs)
+        uart->current_intr = ilog2(uart->pending_intrs);
+}
+
+void u8250_check_ready(u8250_state_t *uart)
+{
+    if (uart->in_ready)
```

Comment:
@ChinYikMing, how do you think?

Reply:
The `in_ready` variable acts as the emulated bit 0 of LSR register. We do need a variable to control the data availability when UART is reading the LSR register since LSR is a read-only register.